File: examples/chapter08/gmresrest.m
%%
% The following experiments are based on a matrix resulting from
% discretization of a partial differential equation.
maxit = 120;  rtol = 1e-8;
d = 50;
A = d^2*gallery('poisson',d);
n = length(A)
b = ones(n,1);

%%
% We compare unrestarted GMRES with three different thresholds for
% restarting.
rest = [maxit 20 30 40];
for j = 1:4
    [~,~,~,~,rv1] = gmres(A,b,rest(j),rtol,maxit/rest(j));
    semilogy(0:length(rv1)-1,rv1,'.-'), hold on
end
axis tight, title('Convergence of restarted GMRES')    % ignore this line
xlabel('m'), ylabel('residual norm')   % ignore this line

%%
% The ``pure'' curve is the lowest one. All of the other curves agree with
% it until they encounter their first restart. 

File: julia/scripts/chapter08/gmresrest.jl

using MatrixDepot
d = 50;
A = d^2*matrixdepot("poisson",d)
@show n = size(A,1)
b = ones(n);

using IterativeSolvers,Plots
maxit = 120;  rtol = 1e-8;
rest = [maxit,20,40,60]
plot([],[],label="",title="Convergence of restarted GMRES",leg=:bottomleft,
    xaxis=("m"), yaxis=(:log10,"residual norm",[1e-8,100]))
for j = 1:4
    x,hist = gmres(A,b,restart=rest[j],tol=rtol,maxiter=maxit,log=true)
    plot!(hist[:resnorm],label="restart = $(rest[j])")
end
plot!([],[],label="")
